### PR TITLE
Issues 252,253: Browse by title and browse by metadata (author)

### LINKED ISSUE #252: Browse By Title
Create a component to browse items by title. This component will be used in the context of the entire repository, or filtered by a community, collection ([example](http://demo.dspace.org/xmlui/handle/10673/2/browse?type=title)), or a metadata field ([example](http://demo.dspace.org/xmlui/browse?type=subject&value=Coaching)). 

It should show a list of items as results. Reuse the [ObjectCollectionComponent](https://github.com/DSpace/dspace-angular/blob/c19369f0277a638be1e42e86f2dc1b8bec06f968/src/app/shared/object-collection/object-collection.component.ts) for this (Collection meaning a set of objects that can either be shown as a list or a grid in this context, not the DSpaceObject named Collection). It will show slightly different fields compared to dspace 6, but that's the point, we want more consistency across different pages in dspace 7. You can find an example in the [SearchResultsComponent](https://github.com/LotteHofstede/dspace-angular/blob/ebbe2e6a93bc6ee2cb9f8e2fbb327b22e3614a20/src/app/%2Bsearch-page/search-results/search-results.component.html)	

Configure the pagination, and enable the gear drop down menu for results per page and sorting. 

Ensure it shows a loading message while loading, and an error message if an error occurs. Fade in the results as they come in. Take a look at e.g. the [Recent Submissions section on the CollectionPageComponent](https://github.com/DSpace/dspace-angular/blob/c19369f0277a638be1e42e86f2dc1b8bec06f968/src/app/%2Bcollection-page/collection-page.component.html#L41) to figure out how to do so

A button to switch between grid and list mode will be added in a separate task. If it's not here yet, test grid mode adding `view=grid` to the URL.

The alphabetic starts-with selector at the top will be created in a separate task.

Create a BrowseByTitlePageComponent that uses your new BrowseByTitleComponent to browse the entire repository at `/browse/title`.  Use `ItemDataService.findAll()` to get the contents for the page. Watch the route for changes to the pagination, sorting and view mode, and update the params for `findAll()` accordingly. Take a look at [the way it happens for the search page](https://github.com/LotteHofstede/dspace-angular/blob/ebbe2e6a93bc6ee2cb9f8e2fbb327b22e3614a20/src/app/+search-page/search-filters/search-filter/search-filter.service.ts#L76) for an example.

No need to add a link to the page for now. A later task will replace this page component with something that dynamically configures the available browse pages based on what the rest api offers, but in order for you to verify that everything works as expected, and for the people that review your PR it is convenient to temporarily create a more hard-coded browse page. So there is also no need to write tests for this page component. Do test the browse component itself though.

### LINKED ISSUE #253: Browse By Metadata
Create a component to browse by metadata values. ([example](http://demo.dspace.org/xmlui/browse?type=author)). 

It should show a list of metadata values as results. Instead of a table with only one column as is used in dspace 6, we'll render it the same way other lists in dspace 7 are rendered, such as [search results](https://dspace7-demo.atmire.com/search) and [recent submissions](https://dspace7-demo.atmire.com/collections/8e0928a0-047a-4369-8883-12669f32dd64): i.e. `<li>`'s with an `<a class="lead">` for the link. So make use of the existing ObjectlistComponent. and create a MetadatumListElementComponent, to be used inside it. Take a look at e.g. the [ItemListElementComponent](https://github.com/DSpace/dspace-angular/blob/c19369f0277a638be1e42e86f2dc1b8bec06f968/src/app/shared/object-list/item-list-element/item-list-element.component.ts) for an example. The `@renderElementsFor` annotation determines for which objects the component is used. No need to design a grid view. It doesn't make a lot of sense for things as simple as metadata vales.

Configure the pagination, and enable the gear drop down menu for results per page and sorting. 

The alphabetic starts-with selector at the top will be created in a separate task.

Ensure it shows a loading message while loading, and an error message if an error occurs. Fade in the results as they come in. Take a look at e.g. the [Recent Submissions section on the CollectionPageComponent](https://github.com/DSpace/dspace-angular/blob/c19369f0277a638be1e42e86f2dc1b8bec06f968/src/app/%2Bcollection-page/collection-page.component.html#L41) to figure out how to do so

Create a BrowseByAuthorPageComponent that uses your new BrowseByMetadataComponent to browse the entire repository at `/browse/author`. The part that fetches the author values from the Rest API and makes them available in a service is not trivial and will be handled in a separate task. For now, just hardcode an `Observable<RemoteData<PaginatedList<Metadatum>>>` to pass to the object list. The following fragment, likely with a few tweaks, should get you started:
```typescript
const mockValues =  [
    {
        key: 'dc.contributor.author',
        value: 'Smith Jr. Donald',
    },
    ...
] as Metadatum[];

const data$ = Observable.of(new RemoteData(
    false,
    false,
    true,
    undefined,
    new PaginatedList(new PageInfo(), mockValues)
));
```
Watch the route for changes to the pagination and sorting and update the mock data accordingly. Take a look at [the way it happens for the search page](https://github.com/LotteHofstede/dspace-angular/blob/ebbe2e6a93bc6ee2cb9f8e2fbb327b22e3614a20/src/app/+search-page/search-filters/search-filter/search-filter.service.ts#L76) for an example.

No need to add a link to the page for now. A later task will replace this page component with something that dynamically configures the available browse pages based on what the rest api offers, but in order for you to verify that everything works as expected, and for the people that review your PR it is convenient to temporarily create a more hard-coded browse page. So there is also no need to write tests for this page component. Do test the browse component itself though.

### PR DESCRIPTION
**Pull request by Atmire-Kristof**
This PR closes #252 and closes #253
Created Browse-By-Title and Browse-By-Author components on routes “/browse/title” and “/browse/author” respectively.
To achieve this, I created a reusable Browse-By component, which accepts a title, a list of any ListableObject to display, paginationConfig and sortConfig.
To display metadata (e.g. authors for browse-by-author), a new component BrowseByEntryListElementComponent was created. This displays a list of the metadata values in the form of links. The links add a “value” parameter to the current url, containing the metadata value.
This parameter is then used to display items linked to said metadata value. (for example all items linked to a specific author)

### CODE REVIEW

File: src/app/+browse-by/+browse-by-author-page/browse-by-author-page.component.ts [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+
+@Component({
+  selector: 'ds-browse-by-author-page',
+  styleUrls: ['./browse-by-author-page.component.scss'],
+  templateUrl: './browse-by-author-page.component.html'
+})
+/**
+ * Component for browsing (items) by author (dc.contributor.author)
+ */
+export class BrowseByAuthorPageComponent implements OnInit {
```
[the remark this answers is not in the document]
Comment (Atmire-Kristof) [reply src/app/+browse-by/+browse-by-author-page/browse-by-author-page.component.ts]: At the bottom of both issued addressed by this PR, it says there is currently no need to write test cases for BrowseByAuthorComponent and BrowseByTitleComponent.
The empty file is merely a placeholder for when tests need to be added.
If you wish for me to delete them anyway, let me know.

[the remark this answers is not in the document]
Comment (tdonohue) [reply src/app/+browse-by/+browse-by-author-page/browse-by-author-page.component.ts]: @artlowel : Is there a reason why we don't want tests for these page components?  (Sidenote: I notice that, even though Travis approves of this PR, we are decreasing our test coverage by 0.5%, which is not a significant change, but a decrease nonetheless)

If we are not yet creating tests here, it seems odd to me to have empty placeholder files (as the existence of those files implies we should have tests in them).  Is there a followup ticket that will ensure these tests get added in?

[the remark this answers is not in the document]
Comment (artlowel) [reply src/app/+browse-by/+browse-by-author-page/browse-by-author-page.component.ts]: Yes, these pages will be removed by #259, they were only created to have somewhere to put the elements of a browse page while working on and testing #253. The large majority of the code inside them is to configure the pagination (that fact that it takes _that_ much code is something we should do something about, but isn't directly related to this task) 

I don't see any good reason to have empty tests files.

[the remark this answers is not in the document]
Comment (tdonohue) [reply src/app/+browse-by/+browse-by-author-page/browse-by-author-page.component.ts]: Thanks for the clarification, @artlowel.  As these page components are temporary, then I'm OK with us leaving these tests out.  I agree and think we should remove the empty test files in this PR.

**Comment by tdonohue** [comment]
@Atmire-Kristof : This looks to now have a minor merge conflict.  It also looks like we are still missing TypeDocs in the `browse-items-response-parsing-service.ts`.

**Comment by tdonohue** [comment]
@Atmire-Kristof: I've tested this PR, and while the functionality works well overall, I noticed that the ordering of "Browse by Title" seems to be random?  I expected titles to be ordered alphabetically, but I'm seeing them in a rather random order, e.g.

1. Test PowerPoint Document
1. Test Webpage
1. Solution structure of apamin determined by nuclear magnetic resonance and distance geometry
1. Test Excel Document
1. Test PDF Document
1. Science, institutional archives and open access: an overview and a pilot survey on the Italian cancer research institutions
 etc.

The Browse by Author functionality seems to work fine though (and authors are ordered alphabetically)

**Pull request merged**